libs/partners/google-vertexai/langchain_google_vertexai/llms.py

```diff
@@ -0,0 +1,471 @@
+from langchain_core.pydantic_v1 import BaseModel, Field, root_validator
+from vertexai.language_models import (  # type: ignore
+    CodeGenerationModel,
+    TextGenerationModel,
+)
+from vertexai.language_models._language_models import (  # type: ignore
+    TextGenerationResponse,
+)
+from vertexai.preview.generative_models import GenerativeModel, Image  # type: ignore
+from vertexai.preview.language_models import (  # type: ignore
+    CodeGenerationModel as PreviewCodeGenerationModel,
+)
+from vertexai.preview.language_models import (
+    TextGenerationModel as PreviewTextGenerationModel,
+)
+
+from langchain_google_vertexai._utils import (
+    create_retry_decorator,
+    get_client_info,
+    is_codey_model,
+    is_gemini_model,
+)
+
+
+def _completion_with_retry(
+    llm: VertexAI,
+    prompt: List[Union[str, Image]],
+    stream: bool = False,
+    is_gemini: bool = False,
+    run_manager: Optional[CallbackManagerForLLMRun] = None,
+    **kwargs: Any,
+) -> Any:
+    """Use tenacity to retry the completion call."""
+    retry_decorator = create_retry_decorator(run_manager=run_manager)
+
+    @retry_decorator
+    def _completion_with_retry_inner(
+        prompt: List[Union[str, Image]], is_gemini: bool = False, **kwargs: Any
+    ) -> Any:
+        if is_gemini:
+            return llm.client.generate_content(
+                prompt, stream=stream, generation_config=kwargs
+            )
+        else:
+            if stream:
+                return llm.client.predict_streaming(prompt[0], **kwargs)
+            return llm.client.predict(prompt[0], **kwargs)
+
+    return _completion_with_retry_inner(prompt, is_gemini, **kwargs)
+
+
+async def _acompletion_with_retry(
+    llm: VertexAI,
+    prompt: str,
+    is_gemini: bool = False,
+    run_manager: Optional[AsyncCallbackManagerForLLMRun] = None,
+    **kwargs: Any,
+) -> Any:
+    """Use tenacity to retry the completion call."""
+    retry_decorator = create_retry_decorator(run_manager=run_manager)
+
+    @retry_decorator
+    async def _acompletion_with_retry_inner(
+        prompt: str, is_gemini: bool = False, **kwargs: Any
+    ) -> Any:
+        if is_gemini:
+            return await llm.client.generate_content_async(
+                prompt, generation_config=kwargs
+            )
+        return await llm.client.predict_async(prompt, **kwargs)
+
+    return await _acompletion_with_retry_inner(prompt, is_gemini, **kwargs)
+
+
+class _VertexAIBase(BaseModel):
+    project: Optional[str] = None
+    "The default GCP project to use when making Vertex API calls."
+    location: str = "us-central1"
+    "The default location to use when making API calls."
+    request_parallelism: int = 5
+    "The amount of parallelism allowed for requests issued to VertexAI models. "
+    "Default is 5."
+    max_retries: int = 6
+    """The maximum number of retries to make when generating."""
+    task_executor: ClassVar[Optional[Executor]] = Field(default=None, exclude=True)
```

Comment:
don't have to do now to keep migration straightforward, but could be good to move this to only be in embeddings if it's only used there in 0.2